Compiler error, no matching constructor

**Minimal example**: Compile the following code with Clang++ with the clad plugin:

```c++
#include "clad/Differentiator/Differentiator.h"

struct T {
  T(int){}
};

double f(double x){
  T t{0};
  return x;
}

int main() {
  double derivative;
  auto f_dx = clad::gradient(f, "x");
  f_dx.execute(3, &derivative);
}
```

**Expected behavior**: Code compiles.

**Observed behavior**: Compiler error:

```
s.cpp:8:8: error: no matching constructor for initialization of 'T'
double f(double x){
       ^
s.cpp:5:3: note: candidate constructor not viable: requires 1 argument, but 0 were provided
  T(int){}
  ^
s.cpp:4:8: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided
struct T {
       ^
s.cpp:4:8: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 0 were provided
1 error generated.

```

**Setup**: Clad built from the GitHub master branch at commit d7e5434c on Ubuntu 22.04 with Clang 11.

--- a/include/clad/Differentiator/ReverseModeVisitor.h
+++ b/include/clad/Differentiator/ReverseModeVisitor.h
@@ -700,34 +700,15 @@ namespace clad {
 
     void PopSwitchStmtInfo() { m_SwitchStmtsData.pop_back(); }
 
-    struct ConstructorPullbackCallInfo {
-      clang::CallExpr* pullbackCE = nullptr;
-      size_t thisAdjointArgIdx = std::numeric_limits<size_t>::max();
-      void updateDThisParm(clang::Expr* dThisE) const;
-      ConstructorPullbackCallInfo() = default;
-      ConstructorPullbackCallInfo(clang::CallExpr* pPullbackCE,
-                                  size_t pThisAdjointArgIdx)
-          : pullbackCE(pPullbackCE), thisAdjointArgIdx(pThisAdjointArgIdx) {}
-
-      bool empty() const { return !pullbackCE; }
-    };
-
-    void setConstructorPullbackCallInfo(clang::CallExpr* pullbackCE,
-                                        size_t thisAdjointArgIdx) {
-      m_ConstructorPullbackCallInfo = {pullbackCE, thisAdjointArgIdx};
-    }
-
-    ConstructorPullbackCallInfo getConstructorPullbackCallInfo() {
-      return m_ConstructorPullbackCallInfo;
-    }
-
-    void resetConstructorPullbackCallInfo() {
-      m_ConstructorPullbackCallInfo = ConstructorPullbackCallInfo{};
-    }
-
   private:
-    ConstructorPullbackCallInfo m_ConstructorPullbackCallInfo;
-    bool m_TrackConstructorPullbackInfo = false;
+    // FIXME: This variable is used to track
+    // whether we're currently visiting an init of a var decl.
+    // This is only necessary because we don't create constructors
+    // explicitly, instead we create a ParenListExpr and expect clang to
+    // build the constructor. However, this only works as var decl inits. In
+    // other cases, we have to use InitListExpr and change the constructor
+    // style. Remove this once we generate constructors explicitly.
+    bool m_TrackVarDeclConstructor = false;
   };
 } // end namespace clad
 

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -2561,13 +2561,10 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     if (isPointerType && VD->getInit() && isa<CXXNewExpr>(VD->getInit()))
       isInitializedByNewExpr = true;
 
-    ConstructorPullbackCallInfo constructorPullbackInfo;
-
     bool isConstructInit =
         VD->getInit() && isa<CXXConstructExpr>(VD->getInit()->IgnoreImplicit());
     const CXXRecordDecl* RD = VD->getType()->getAsCXXRecordDecl();
     bool isNonAggrClass = RD && !RD->isAggregate();
-    bool emptyInitListInit = isNonAggrClass;
 
     // VDDerivedInit now serves two purposes -- as the initial derivative value
     // or the size of the derivative array -- depending on the primal type.
@@ -2598,18 +2595,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
         VDDerivedInit = getZeroInit(VDDerivedType);
       else
         VDDerivedInit = initDiff.getExpr_dx();
-    }
-
-    if (isConstructInit) {
-      m_TrackConstructorPullbackInfo = true;
-      initDiff = Visit(VD->getInit());
-      m_TrackConstructorPullbackInfo = false;
-      constructorPullbackInfo = getConstructorPullbackCallInfo();
-      resetConstructorPullbackCallInfo();
-      if (initDiff.getExpr_dx()) {
-        VDDerivedInit = initDiff.getExpr_dx();
-        emptyInitListInit = false;
-      }
     }
 
     // if VD is a pointer type, then the initial value is set to the derived
@@ -2655,8 +2640,11 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
           derivedE = BuildOp(UnaryOperatorKind::UO_Deref, derivedE);
       }
 
-      if (VD->getInit() && !isConstructInit)
+      if (VD->getInit()) {
+        llvm::SaveAndRestore<bool> saveTrackVarDecl(m_TrackVarDeclConstructor,
+                                                    true);
         initDiff = Visit(VD->getInit(), derivedE);
+      }
 
       // If we are differentiating `VarDecl` corresponding to a local variable
       // inside a loop, then we need to reset it to 0 at each iteration.
@@ -2720,49 +2708,52 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
       VDClone = BuildGlobalVarDecl(VDCloneType, VD->getNameAsString(),
                                    initDiff.getExpr(), VD->isDirectInit());
 
-    // We initialize adjoints with original variables as part of
-    // the strategy to maintain the structure of the original variable.
-    // After that, we'll zero-initialize the adjoint. e.g.
-    // ```
-    // std::vector<...> v{x, y, z};
-    // std::vector<...> _d_v{v}; // The length of the vector is preserved
-    // clad::zero_init(_d_v);
-    // ```
-    // Also, if the original is initialized with a zero-constructor, it can be
-    // used for the adjoint as well.
-    bool shouldCopyInitialize =
-        isConstructInit && emptyInitListInit &&
-        cast<CXXConstructExpr>(VD->getInit()->IgnoreImplicit())->getNumArgs();
-    shouldCopyInitialize = shouldCopyInitialize &&
-                           (VDType->getAsCXXRecordDecl() &&
-                            utils::isCopyable(VDType->getAsCXXRecordDecl()));
-    if (shouldCopyInitialize) {
-      Expr* copyExpr = BuildDeclRef(VDClone);
-      QualType origTy = VDClone->getType();
-      if (isInsideLoop) {
-        StmtDiff pushPop = StoreAndRestore(
-            BuildDeclRef(VDDerived), /*prefix=*/"_t", /*moveToTape=*/true);
-        addToCurrentBlock(pushPop.getStmt(), direction::forward);
-        addToCurrentBlock(pushPop.getStmt_dx(), direction::reverse);
+    if (isConstructInit) {
+      // We initialize adjoints with original variables as part of
+      // the strategy to maintain the structure of the original variable.
+      // After that, we'll zero-initialize the adjoint. e.g.
+      // ```
+      // std::vector<...> v{x, y, z};
+      // std::vector<...> _d_v{v}; // The length of the vector is preserved
+      // clad::zero_init(_d_v);
+      // ```
+      // Also, if the original is initialized with a zero-constructor, it can be
+      // used for the adjoint as well.
+      bool shouldCopyInitialize =
+          isNonAggrClass &&
+          cast<CXXConstructExpr>(VD->getInit()->IgnoreImplicit())
+              ->getNumArgs() &&
+          utils::isCopyable(VDType->getAsCXXRecordDecl());
+      if (initDiff.getStmt_dx()) {
+        SetDeclInit(VDDerived, initDiff.getExpr_dx());
+      } else if (shouldCopyInitialize) {
+        Expr* copyExpr = BuildDeclRef(VDClone);
+        QualType origTy = VDClone->getType();
+        if (isInsideLoop) {
+          StmtDiff pushPop = StoreAndRestore(
+              BuildDeclRef(VDDerived), /*prefix=*/"_t", /*moveToTape=*/true);
+          addToCurrentBlock(pushPop.getStmt(), direction::forward);
+          addToCurrentBlock(pushPop.getStmt_dx(), direction::reverse);
+        }
+        // if VDClone is volatile, we have to use const_cast to be able to use
+        // most copy constructors.
+        if (origTy.isVolatileQualified()) {
+          Qualifiers quals(origTy.getQualifiers());
+          quals.removeVolatile();
+          QualType castTy = m_Sema.BuildQualifiedType(
+              origTy.getUnqualifiedType(), noLoc, quals);
+          castTy = m_Context.getLValueReferenceType(castTy);
+          SourceRange range = utils::GetValidSRange(m_Sema);
+          copyExpr =
+              m_Sema
+                  .BuildCXXNamedCast(noLoc, tok::kw_const_cast,
+                                     m_Context.getTrivialTypeSourceInfo(
+                                         castTy, utils::GetValidSLoc(m_Sema)),
+                                     copyExpr, range, range)
+                  .get();
+        }
+        SetDeclInit(VDDerived, copyExpr, /*DirectInit=*/true);
       }
-      // if VDClone is volatile, we have to use const_cast to be able to use
-      // most copy constructors.
-      if (origTy.isVolatileQualified()) {
-        Qualifiers quals(origTy.getQualifiers());
-        quals.removeVolatile();
-        QualType castTy = m_Sema.BuildQualifiedType(origTy.getUnqualifiedType(),
-                                                    noLoc, quals);
-        castTy = m_Context.getLValueReferenceType(castTy);
-        SourceRange range = utils::GetValidSRange(m_Sema);
-        copyExpr =
-            m_Sema
-                .BuildCXXNamedCast(noLoc, tok::kw_const_cast,
-                                   m_Context.getTrivialTypeSourceInfo(
-                                       castTy, utils::GetValidSLoc(m_Sema)),
-                                   copyExpr, range, range)
-                .get();
-      }
-      SetDeclInit(VDDerived, copyExpr, /*DirectInit=*/true);
     }
 
     if (isPointerType && derivedVDE) {
@@ -2810,11 +2801,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
          VDType != VDClone->getType()))
       m_DeclReplacements[VD] = VDClone;
 
-    if (!constructorPullbackInfo.empty()) {
-      Expr* dThisE =
-          BuildOp(UnaryOperatorKind::UO_AddrOf, BuildDeclRef(VDDerived));
-      constructorPullbackInfo.updateDThisParm(dThisE);
-    }
     return DeclDiff<VarDecl>(VDClone, VDDerived);
   }
 
@@ -4081,8 +4067,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     }
 
     const CXXRecordDecl* RD = CD->getParent();
-    QualType recordType = m_Context.getRecordType(RD);
-    QualType recordPointerType = m_Context.getPointerType(recordType);
 
     // FIXME: consider moving non-diff analysis to DiffPlanner.
     bool nonDiff = clad::utils::hasNonDifferentiableAttribute(CE);
@@ -4093,7 +4077,7 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     // If the callee function takes arguments by reference then it can affect
     // derivatives even if there is no `dfdx()` and thus we should call the
     // derived function.
-    if (!nonDiff && !(dfdx() || m_TrackConstructorPullbackInfo))
+    if (!nonDiff && !dfdx())
       nonDiff = true;
 
     // If all arguments are constant literals, then this does not contribute to
@@ -4111,18 +4095,8 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     if (!nonDiff) {
       // Try to create a pullback constructor call
       llvm::SmallVector<Expr*, 4> pullbackArgs;
-      // dThisE = adjoint of the object being created by this constructor call.
-      //
-      // We cannot fill this arg yet because the object has not yet been
-      // created. The caller which triggers 'VisitCXXConstructExpr' is
-      // responsible for updating this arg.
-      Expr* dThisE = nullptr;
-      if (m_TrackConstructorPullbackInfo)
-        dThisE = getZeroInit(recordPointerType);
-      else
-        dThisE = BuildOp(UnaryOperatorKind::UO_AddrOf, dfdx(),
-                         m_DiffReq->getLocation());
-
+      Expr* dThisE = BuildOp(UnaryOperatorKind::UO_AddrOf, dfdx(),
+                             m_DiffReq->getLocation());
       pullbackArgs.append(primalArgs.begin(), primalArgs.end());
       pullbackArgs.push_back(dThisE);
       pullbackArgs.append(adjointArgs.begin(), adjointArgs.end());
@@ -4137,10 +4111,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
               m_Builder.BuildCallToCustomDerivativeOrNumericalDiff(
                   customPullbackName, pullbackArgs, getCurrentScope(), CE)) {
         curRevBlock.insert(it, customPullbackCall);
-        if (m_TrackConstructorPullbackInfo) {
-          setConstructorPullbackCallInfo(
-              llvm::cast<CallExpr>(customPullbackCall), primalArgs.size());
-        }
       }
     }
     // FIXME: If no compatible custom constructor pullback is found then try
@@ -4206,7 +4176,7 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
       // ReturnStmt. However, to support member exprs/calls of constructors, we
       // need to explicitly generate a constructor and not rely on higher level
       // Sema functions.
-      if (CE->isListInitialization() || !m_TrackConstructorPullbackInfo) {
+      if (CE->isListInitialization() || !m_TrackVarDeclConstructor) {
         clonedArgsE = m_Sema.ActOnInitList(noLoc, primalArgs, noLoc).get();
       } else {
         if (CE->getNumArgs() == 0) {
@@ -4489,10 +4459,5 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
         m_Builder.BuildCallToCustomDerivativeOrNumericalDiff(
             forwPassFnName, args, getCurrentScope(), callSite);
     return customForwPassCE;
-  }
-
-  void ReverseModeVisitor::ConstructorPullbackCallInfo::updateDThisParm(
-      Expr* dThisE) const {
-    pullbackCE->setArg(thisAdjointArgIdx, dThisE);
   }
 } // end namespace clad

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -2566,6 +2566,30 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     const CXXRecordDecl* RD = VD->getType()->getAsCXXRecordDecl();
     bool isNonAggrClass = RD && !RD->isAggregate();
 
+    // We initialize adjoints with original variables as part of
+    // the strategy to maintain the structure of the original variable.
+    // After that, we'll zero-initialize the adjoint. e.g.
+    // ```
+    // std::vector<...> v{x, y, z};
+    // std::vector<...> _d_v{v}; // The length of the vector is preserved
+    // clad::zero_init(_d_v);
+    // ```
+    // Also, if the original is initialized with a zero-constructor, it can be
+    // used for the adjoint as well.
+    bool shouldCopyInitialize =
+        isConstructInit && isNonAggrClass &&
+        cast<CXXConstructExpr>(VD->getInit()->IgnoreImplicit())->getNumArgs() &&
+        utils::isCopyable(VDType->getAsCXXRecordDecl());
+
+    // Temporarily initialize the object with `*nullptr` to avoid
+    // a potential error because of non-existing default constructor.
+    if (!VDDerivedInit && shouldCopyInitialize) {
+      QualType ptrType =
+          m_Context.getPointerType(VDDerivedType.getUnqualifiedType());
+      Expr* dummy = getZeroInit(ptrType);
+      VDDerivedInit = BuildOp(UO_Deref, dummy);
+    }
+
     // VDDerivedInit now serves two purposes -- as the initial derivative value
     // or the size of the derivative array -- depending on the primal type.
     if (promoteToFnScope)
@@ -2709,21 +2733,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
                                    initDiff.getExpr(), VD->isDirectInit());
 
     if (isConstructInit) {
-      // We initialize adjoints with original variables as part of
-      // the strategy to maintain the structure of the original variable.
-      // After that, we'll zero-initialize the adjoint. e.g.
-      // ```
-      // std::vector<...> v{x, y, z};
-      // std::vector<...> _d_v{v}; // The length of the vector is preserved
-      // clad::zero_init(_d_v);
-      // ```
-      // Also, if the original is initialized with a zero-constructor, it can be
-      // used for the adjoint as well.
-      bool shouldCopyInitialize =
-          isNonAggrClass &&
-          cast<CXXConstructExpr>(VD->getInit()->IgnoreImplicit())
-              ->getNumArgs() &&
-          utils::isCopyable(VDType->getAsCXXRecordDecl());
       if (initDiff.getStmt_dx()) {
         SetDeclInit(VDDerived, initDiff.getExpr_dx());
       } else if (shouldCopyInitialize) {

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -2551,7 +2551,8 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     VarDecl* VDDerived = nullptr;
     bool isPointerType = VDType->isPointerType();
     bool isInitializedByNewExpr = false;
-    bool initializeDerivedVar = true;
+    bool initializeDerivedVar = m_DiffReq.shouldHaveAdjoint(VD) &&
+                                !clad::utils::hasNonDifferentiableAttribute(VD);
 
     if (Expr* size = getStdInitListSizeExpr(VD->getInit()))
       VDDerivedInit = size;
@@ -2641,9 +2642,6 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
       VDDerived = BuildGlobalVarDecl(
           VDDerivedType, "_d_" + VD->getNameAsString(), VDDerivedInit);
 
-    if (!m_DiffReq.shouldHaveAdjoint(VD))
-      VDDerived = nullptr;
-
     // If `VD` is a reference to a local variable, then it is already
     // differentiated and should not be differentiated again.
     // If `VD` is a reference to a non-local variable then also there's no
@@ -2651,7 +2649,7 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     if (!isRefType && (!isPointerType || isInitializedByNewExpr)) {
       Expr* derivedE = nullptr;
 
-      if (VDDerived && !clad::utils::hasNonDifferentiableAttribute(VD)) {
+      if (VDDerived) {
         derivedE = BuildDeclRef(VDDerived);
         if (isInitializedByNewExpr)
           derivedE = BuildOp(UnaryOperatorKind::UO_Deref, derivedE);
@@ -2688,7 +2686,7 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
 
     VarDecl* VDClone = nullptr;
     Expr* derivedVDE = nullptr;
-    if (VDDerived && m_DiffReq.shouldHaveAdjoint(const_cast<VarDecl*>(VD)))
+    if (VDDerived)
       derivedVDE = BuildDeclRef(VDDerived);
     // FIXME: Add extra parantheses if derived variable pointer is pointing to a
     // class type object.

--- a/test/Gradient/NonDifferentiable.C
+++ b/test/Gradient/NonDifferentiable.C
@@ -180,9 +180,7 @@ int main() {
     // CHECK-NEXT: }
 
     // CHECK: void fn_non_diff_var_grad(double i, double j, double *_d_i, double *_d_j) {
-    // CHECK-NEXT:     double _d_k = 0.;
     // CHECK-NEXT:     double k = i * i * j;
-    // CHECK-NEXT:     _d_k += 1;
     // CHECK-NEXT: }
     
     // CHECK: void mem_fn_1_pullback(double i, double j, double _d_y, SimpleFunctions1 *_d_this, double *_d_i, double *_d_j) {

--- a/test/Gradient/UserDefinedTypes.C
+++ b/test/Gradient/UserDefinedTypes.C
@@ -777,6 +777,22 @@ double fn22(double x, double y) {
 // CHECK-NEXT:      }
 // CHECK-NEXT:  }
 
+struct StructNoDefConstr {
+  StructNoDefConstr(int){}
+};
+
+double fn23(double x){
+  StructNoDefConstr t{0};
+  return x;
+}
+
+// CHECK:  void fn23_grad(double x, double *_d_x) {
+// CHECK-NEXT:      StructNoDefConstr t(0);
+// CHECK-NEXT:      StructNoDefConstr _d_t(t);
+// CHECK-NEXT:      clad::zero_init(_d_t);
+// CHECK-NEXT:      *_d_x += 1;
+// CHECK-NEXT:  }
+
 void print(const Tangent& t) {
   for (int i = 0; i < 5; ++i) {
     printf("%.2f", t.data[i]);
@@ -873,6 +889,8 @@ int main() {
 
     INIT_GRADIENT(fn22);
     TEST_GRADIENT(fn22, /*numOfDerivativeArgs=*/2, 3, 2, &d_i, &d_j);    // CHECK-EXEC: {8.00, 0.00}
+
+    INIT_GRADIENT(fn23);
 }
 
 // CHECK: void someMemFn2_pullback(double i, double j, double _d_y, Tangent *_d_this, double *_d_i, double *_d_j) const {

